src/language/core.md

```diff
@@ -0,0 +1,152 @@
+# The core of hacspec
+
+## Crates and modules
+
+hacspec only supports single-module crates, due to a technical limitation
+of the Rust compiler. Inside this single file, a hacspec program shall always
+start with:
+
+```rust, noplaypen
+use hacspec_lib::*;
+
+// Optional: dependencies on other crates containing hacspec programs
+use foo_dep::*;
+```
+
+No other form of `use` is allowed in hacspec, because allowing Rust's
+complex import patterns would increase the complexity of the hacspec compiler
+and conflict with the module systems of most proof assistants.
+
+## Functions
+
+hacspec is a functional language, and only supports the declaration of
+top-level functions:
+
+```rust, noplaypen
+fn foo(x: bool) -> () {
+    ...
+}
+```
+
+The functions can take any number of arguments, and may return a value (or not).
+Note that recursive functions are forbidden in hacspec.
+
+## Basic Types
+
+hacspec supports all the Rust primitive types: integers (signed and unsigned),
+booleans, unit, tuples.
+
+Type aliases are allowed in Hacspec:
+
+```rust, noplaypen
+type FooAlias = u32;
+```
+
+## Borrows
+
+hacspec forbids mutable borrows at all places. Immutable borrows are allowed
+in hacspec, but only for function arguments. Indeed, you can declare a function
+argument as immutably borrowed:
+
+```rust, noplaypen
+fn foo(arg: &Seq<u8>) {
+    ...
+}
+```
+
+You can also immutably borrow a value at the call site of a function:
+
+```rust, noplaypen
+foo(&Seq::<u8>::new(64))
+```
+
+In particular, return types cannot contain references, and the same is true
+for types inside tuples or any data structures.
+
+## Constants
+
+hacspec allows the declaration of constants:
+
+```rust, noplaypen
+const FOO_BAR : bool = false;
+```
+
+## Assignments
+
+Inside a function body, hacspec allows regular Rust let-bindings:
+
+```rust, noplaypen
+let x = ...;
+```
+
+hacspec also allows mutable let bindings, and subsequent reassignments:
+
+```rust, noplaypen
+let mut x = ...;
+...
+x = ...;
+```
+
+This allowing of mutable variable might come as a contradiction to hacspec's
+claim of forbidding mutable state. But in fact, mutable local variables in
```

Comment:
Did we talk about not allowing mutable state before?

Reply:
Yes, in the main readme